File: src/templates/nextjs.ts
import { NextjsConfig, DbProvider, OrmChoice } from "../types";

export const buildPackageJson = (cfg: NextjsConfig): string => {
  const hasOrm = cfg.orm !== "none";

  const deps: Record<string, string> = {
    next: "15.1.0",
    react: "19.0.0",
    "react-dom": "19.0.0",
  };

  if (cfg.nextAuth) {
    deps["next-auth"] = "5.0.0-beta.25";
  }

  if (cfg.orm === "prisma") {
    deps["@prisma/client"] = "6.0.1";
  }

  if (cfg.orm === "drizzle") {
    deps["drizzle-orm"] = "0.38.2";
    if (cfg.dbProvider === "postgres") deps["postgres"] = "3.4.5";
    if (cfg.dbProvider === "mysql") deps["mysql2"] = "3.12.0";
    if (cfg.dbProvider === "sqlite") deps["better-sqlite3"] = "11.7.0";
  }

  const devDeps: Record<string, string> = {
    typescript: "5.7.2",
    "@types/node": "22.10.2",
    "@types/react": "19.0.1",
    "@types/react-dom": "19.0.2",
    eslint: "9.17.0",
    "eslint-config-next": "15.1.0",
  };

  if (cfg.tailwind) {
    devDeps["tailwindcss"] = "3.4.16";
    devDeps["postcss"] = "8.4.49";
    devDeps["autoprefixer"] = "10.4.20";
  }

  if (cfg.orm === "prisma") devDeps["prisma"] = "6.0.1";
  if (cfg.orm === "drizzle") devDeps["drizzle-kit"] = "0.30.1";

  if (cfg.jest) {
    devDeps["jest"] = "29.7.0";
    devDeps["jest-environment-jsdom"] = "29.7.0";
    devDeps["@testing-library/react"] = "16.1.0";
    devDeps["@testing-library/jest-dom"] = "6.6.3";
    devDeps["@types/jest"] = "29.5.14";
    devDeps["ts-jest"] = "29.2.5";
  }

  const scripts: Record<string, string> = {
    dev: "next dev",
    build: "next build",
    start: "next start",
    lint: "next lint",
  };

  if (hasOrm) {
    if (cfg.orm === "prisma") {
      scripts["db:push"] = "prisma db push";
      scripts["db:studio"] = "prisma studio";
    } else {
      scripts["db:push"] = "drizzle-kit push";
      scripts["db:studio"] = "drizzle-kit studio";
    }
  }

  if (cfg.jest) {
    scripts["test"] = "jest";
    scripts["test:watch"] = "jest --watch";
  }

  return JSON.stringify(
    {
      name: "my-app",
      version: "0.1.0",
      private: true,
      scripts,
      dependencies: deps,
      devDependencies: devDeps,
    },
    null,
    2
  );
};

export const buildTsConfig = (): string => `{
  "compilerOptions": {
    "target": "ES2017",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "bundler",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true,
    "plugins": [{ "name": "next" }],
    "paths": {
      "@/*": ["./src/*"]
    }
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
  "exclude": ["node_modules"]
}`;

export const buildLayoutTsx = (cfg: NextjsConfig): string => `import type { Metadata } from "next";
import "./globals.css";

export const metadata: Metadata = {
  title: "Create Ruxum App",
  description: "Scaffolded with create-ruxum-app",
};

export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  return (
    <html lang="en">
      <body className="antialiased">{children}</body>
    </html>
  );
}
`;

export const buildPageTsx = (): string => `export default function Home() {
  return (
    <main className="flex min-h-screen flex-col items-center justify-center bg-gradient-to-b from-[#2e026d] to-[#15162c] text-white">
      <div className="container flex flex-col items-center justify-center gap-12 px-4 py-16">
        <h1 className="text-5xl font-extrabold tracking-tight sm:text-[5rem]">
          Create <span className="text-[hsl(280,100%,70%)]">Ruxum</span> App
        </h1>
        <div className="grid grid-cols-1 gap-4 sm:grid-cols-2 md:gap-8">
          <div className="flex max-w-xs flex-col gap-4 rounded-xl bg-white/10 p-4 hover:bg-white/20">
            <h3 className="text-2xl font-bold">Documentation →</h3>
            <div className="text-lg">
              Learn more about Ruxum and how to build high-performance APIs.
            </div>
          </div>
          <div className="flex max-w-xs flex-col gap-4 rounded-xl bg-white/10 p-4 hover:bg-white/20">
            <h3 className="text-2xl font-bold">Next.js Frontend →</h3>
            <div className="text-lg">
              Explore the Next.js frontend and how it integrates with Axum.
            </div>
          </div>
        </div>
      </div>
    </main>
  );
}
`;

export const buildNextGitignore = (): string => `# Next.js
.next/
out/
build/

# production
dist/

# misc
.DS_Store
*.pem

# debug
npm-debug.log*
yarn-debug.log*
yarn-error.log*

# local env files
.env*.local

# vercel
.vercel

# typescript
*.tsbuildinfo
next-env.d.ts
node_modules/
`;

export const buildNextEnvExample = (cfg: NextjsConfig): string => {
  let s = "";
  if (cfg.orm !== "none") {
    if (cfg.dbProvider === "sqlite") s += "DATABASE_URL=file:./dev.db\n";
    else s += "DATABASE_URL=postgresql://user:password@localhost:5432/mydb\n";
  }
  if (cfg.nextAuth) {
    s += "NEXTAUTH_SECRET=changeme\nNEXTAUTH_URL=http://localhost:3000\n";
  }
  return s || "# No env vars needed yet\n";
};

File: src/scaffold/nextjs.rs
use anyhow::Result;
use std::fs;
use std::path::Path;

use crate::config::{DbProvider, NextjsConfig, Orm};

pub fn scaffold(dir: &Path, cfg: &NextjsConfig, project_name: &str) -> Result<()> {
    let has_orm = cfg.orm != Orm::None;

    fs::create_dir_all(dir.join("src/app"))?;
    fs::create_dir_all(dir.join("src/lib"))?;
    if cfg.shadcn {
        fs::create_dir_all(dir.join("src/components/ui"))?;
    }
    if has_orm {
        match cfg.orm {
            Orm::Prisma => fs::create_dir_all(dir.join("prisma"))?,
            Orm::Drizzle => fs::create_dir_all(dir.join("drizzle"))?,
            Orm::None => {}
        }
    }
    if cfg.next_auth {
        fs::create_dir_all(dir.join("src/app/api/auth/[...nextauth]"))?;
    }
    if cfg.jest {
        fs::create_dir_all(dir.join("__tests__"))?;
    }

    write(dir, "package.json", &package_json(cfg, project_name))?;
    write(dir, "tsconfig.json", tsconfig_json())?;
    write(dir, "next.config.mjs", &next_config(cfg))?;
    write(dir, "src/app/layout.tsx", &layout_tsx(cfg, project_name))?;
    write(dir, "src/app/page.tsx", page_tsx())?;
    write(dir, ".env.example", &env_example(cfg))?;
    write(dir, ".gitignore", nextjs_gitignore())?;
    write(dir, "eslint.config.mjs", eslint_config())?;

    if cfg.tailwind {
        write(dir, "src/app/globals.css", &globals_css(cfg))?;
    } else {
        write(dir, "src/app/globals.css", globals_css_plain())?;
    }

    if cfg.shadcn {
        write(dir, "components.json", &components_json(cfg))?;
    }

    if has_orm {
        write(dir, "src/lib/db.ts", &db_ts(cfg))?;
        match cfg.orm {
            Orm::Prisma => write(dir, "prisma/schema.prisma", &prisma_schema(cfg))?,
            Orm::Drizzle => {
                write(dir, "drizzle/schema.ts", &drizzle_schema(cfg))?;
                write(dir, "drizzle.config.ts", &drizzle_config(cfg))?;
            }
            Orm::None => {}
        }
    }

    if cfg.next_auth {
        write(dir, "src/lib/auth.ts", auth_ts())?;
        write(
            dir,
            "src/app/api/auth/[...nextauth]/route.ts",
            nextauth_route(),
        )?;
    }

    if cfg.jest {
        write(dir, "jest.config.ts", jest_config())?;
        write(dir, "jest.setup.ts", jest_setup())?;
        write(dir, "__tests__/page.test.tsx", page_test())?;
    }

    Ok(())
}

fn write(dir: &Path, path: &str, content: &str) -> Result<()> {
    let full = dir.join(path);
    if let Some(parent) = full.parent() {
        fs::create_dir_all(parent)?;
    }
    fs::write(full, content)?;
    Ok(())
}

fn package_json(cfg: &NextjsConfig, project_name: &str) -> String {
    let template = include_str!("../templates/nextjs/package_json.rs");
    let has_orm = cfg.orm != Orm::None;

    let mut deps: Vec<(&str, &str)> = vec![
        ("next", "16.2.1"),
        ("react", "19.2.4"),
        ("react-dom", "19.2.4"),
    ];

    if cfg.next_auth {
        deps.push(("next-auth", "5.0.0-beta.29"));
    }

    if cfg.orm == Orm::Prisma {
        deps.push(("@prisma/client", "6.9.0"));
    }

    if cfg.orm == Orm::Drizzle {
        match cfg.db_provider {
            DbProvider::Postgres => {
                deps.push(("drizzle-orm", "0.44.0"));
                deps.push(("postgres", "3.4.5"));
            }
            DbProvider::Mysql => {
                deps.push(("drizzle-orm", "0.44.0"));
                deps.push(("mysql2", "3.14.0"));
            }
            DbProvider::Sqlite => {
                deps.push(("drizzle-orm", "0.44.0"));
                deps.push(("better-sqlite3", "11.10.0"));
            }
        }
    }

    let mut dev_deps: Vec<(&str, &str)> = vec![
        ("typescript", "6.0.2"),
        ("@types/node", "25.5.0"),
        ("@types/react", "19.2.14"),
        ("@types/react-dom", "19.2.7"),
        ("eslint", "9.30.0"),
        ("eslint-config-next", "16.2.1"),
    ];

    if cfg.tailwind {
        dev_deps.push(("tailwindcss", "4.2.2"));
        dev_deps.push(("@tailwindcss/postcss", "4.2.2"));
        dev_deps.push(("postcss", "8.5.8"));
        if cfg.shadcn {
            dev_deps.push(("@tailwindcss/typography", "0.5.19"));
        }
    }

    if cfg.orm == Orm::Prisma {
        dev_deps.push(("prisma", "6.9.0"));
    }

    if cfg.orm == Orm::Drizzle {
        dev_deps.push(("drizzle-kit", "0.31.0"));
        if cfg.db_provider == DbProvider::Sqlite {
            dev_deps.push(("@types/better-sqlite3", "7.6.13"));
        }
    }

    if cfg.jest {
        dev_deps.push(("jest", "29.7.0"));
        dev_deps.push(("jest-environment-jsdom", "29.7.0"));
        dev_deps.push(("@testing-library/react", "16.3.0"));
        dev_deps.push(("@testing-library/jest-dom", "6.6.3"));
        dev_deps.push(("@types/jest", "29.5.14"));
        dev_deps.push(("ts-jest", "29.4.0"));
    }

    let deps_str = deps
        .iter()
        .map(|(k, v)| format!("    \"{k}\": \"{v}\""))
        .collect::<Vec<_>>()
        .join(",\n");

    let dev_deps_str = dev_deps
        .iter()
        .map(|(k, v)| format!("    \"{k}\": \"{v}\""))
        .collect::<Vec<_>>()
        .join(",\n");

    let mut scripts = vec![
        ("dev", "next dev"),
        ("build", "next build"),
        ("start", "next start"),
        ("lint", "next lint"),
    ];
    if has_orm && cfg.orm == Orm::Prisma {
        scripts.push(("db:push", "prisma db push"));
        scripts.push(("db:studio", "prisma studio"));
    }
    if has_orm && cfg.orm == Orm::Drizzle {
        scripts.push(("db:push", "drizzle-kit push"));
        scripts.push(("db:studio", "drizzle-kit studio"));
    }
    if cfg.jest {
        scripts.push(("test", "jest"));
        scripts.push(("test:watch", "jest --watch"));
    }

    let scripts_str = scripts
        .iter()
        .map(|(k, v)| format!("    \"{k}\": \"{v}\""))
        .collect::<Vec<_>>()
        .join(",\n");

    template
        .replace("{{project_name}}", project_name)
        .replace("{{scripts}}", &scripts_str)
        .replace("{{dependencies}}", &deps_str)
        .replace("{{devDependencies}}", &dev_deps_str)
}

fn tsconfig_json() -> &'static str {
    include_str!("../templates/nextjs/tsconfig_json.rs")
}

fn next_config(_cfg: &NextjsConfig) -> &'static str {
    include_str!("../templates/nextjs/next_config.rs")
}

fn layout_tsx(cfg: &NextjsConfig, project_name: &str) -> String {
    let template = include_str!("../templates/nextjs/layout_tsx.rs");
    let css_import = if cfg.tailwind {
        "import \"./globals.css\";"
    } else {
        "import \"./globals.css\";"
    };

    template
        .replace("{{project_name}}", project_name)
        .replace("{{css_import}}", css_import)
}

fn page_tsx() -> &'static str {
    include_str!("../templates/nextjs/page_tsx.rs")
}

fn globals_css(cfg: &NextjsConfig) -> String {
    let template = include_str!("../templates/nextjs/globals_css.rs");
    let typography = if cfg.shadcn {
        "\n@plugin \"@tailwindcss/typography\";"
    } else {
        ""
    };

    template.replace("{{typography}}", typography)
}

fn globals_css_plain() -> &'static str {
    "* {\n  box-sizing: border-box;\n  margin: 0;\n  padding: 0;\n}\n"
}

fn components_json(_cfg: &NextjsConfig) -> String {
    r#"{
  "$schema": "https://ui.shadcn.com/schema.json",
  "style": "default",
  "rsc": true,
  "tsx": true,
  "tailwind": {
    "config": "",
    "css": "src/app/globals.css",
    "baseColor": "neutral",
    "cssVariables": true
  },
  "aliases": {
    "components": "@/components",
    "utils": "@/lib/utils",
    "ui": "@/components/ui",
    "lib": "@/lib",
    "hooks": "@/hooks"
  },
  "iconLibrary": "lucide"
}
"#
    .to_string()
}

fn db_ts(cfg: &NextjsConfig) -> String {
    match cfg.orm {
        Orm::Prisma => r#"import { PrismaClient } from "@prisma/client";

const globalForPrisma = globalThis as unknown as { prisma: PrismaClient };

export const db =
  globalForPrisma.prisma ||
  new PrismaClient({
    log: process.env.NODE_ENV === "development" ? ["query"] : [],
  });

if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = db;
"#
        .to_string(),
        Orm::Drizzle => match cfg.db_provider {
            DbProvider::Postgres => r#"import { drizzle } from "drizzle-orm/postgres-js";
import postgres from "postgres";
import * as schema from "../../drizzle/schema";

const client = postgres(process.env.DATABASE_URL!);
export const db = drizzle(client, { schema });
"#
            .to_string(),
            DbProvider::Mysql => r#"import { drizzle } from "drizzle-orm/mysql2";
import mysql from "mysql2/promise";
import * as schema from "../../drizzle/schema";

const client = await mysql.createConnection(process.env.DATABASE_URL!);
export const db = drizzle(client, { schema, mode: "default" });
"#
            .to_string(),
            DbProvider::Sqlite => r#"import { drizzle } from "drizzle-orm/better-sqlite3";
import Database from "better-sqlite3";
import * as schema from "../../drizzle/schema";

const client = new Database(process.env.DATABASE_URL ?? "dev.db");
export const db = drizzle(client, { schema });
"#
            .to_string(),
        },
        Orm::None => String::new(),
    }
}

fn prisma_schema(cfg: &NextjsConfig) -> String {
    let template = include_str!("../templates/nextjs/prisma_schema.rs");
    let provider = match cfg.db_provider {
        DbProvider::Postgres => "postgresql",
        DbProvider::Mysql => "mysql",
        DbProvider::Sqlite => "sqlite",
    };

    template.replace("{{provider}}", provider)
}

fn drizzle_schema(cfg: &NextjsConfig) -> String {
    match cfg.db_provider {
        DbProvider::Postgres => r#"import { pgTable, serial, text, timestamp } from "drizzle-orm/pg-core";

// Add your tables here
// export const users = pgTable("users", {
//   id: serial("id").primaryKey(),
//   email: text("email").notNull().unique(),
//   createdAt: timestamp("created_at").defaultNow().notNull(),
// });
"#
        .to_string(),
        DbProvider::Mysql => r#"import { mysqlTable, serial, text, timestamp } from "drizzle-orm/mysql-core";

// Add your tables here
// export const users = mysqlTable("users", {
//   id: serial("id").primaryKey(),
//   email: text("email").notNull().unique(),
//   createdAt: timestamp("created_at").defaultNow().notNull(),
// });
"#
        .to_string(),
        DbProvider::Sqlite => r#"import { sqliteTable, integer, text } from "drizzle-orm/sqlite-core";

// Add your tables here
// export const users = sqliteTable("users", {
//   id: integer("id").primaryKey({ autoIncrement: true }),
//   email: text("email").notNull().unique(),
// });
"#
        .to_string(),
    }
}

fn drizzle_config(cfg: &NextjsConfig) -> String {
    let template = include_str!("../templates/nextjs/drizzle_config.rs");
    let dialect = match cfg.db_provider {
        DbProvider::Postgres => "postgresql",
        DbProvider::Mysql => "mysql",
        DbProvider::Sqlite => "sqlite",
    };

    template.replace("{{dialect}}", dialect)
}

fn auth_ts() -> &'static str {
    include_str!("../templates/nextjs/auth_config.rs")
}

fn nextauth_route() -> &'static str {
    r#"import { handlers } from "@/lib/auth";

export const { GET, POST } = handlers;
"#
}

fn jest_config() -> &'static str {
    include_str!("../templates/nextjs/jest_config.rs")
}

fn jest_setup() -> &'static str {
    r#"import "@testing-library/jest-dom";
"#
}

fn page_test() -> &'static str {
    r#"import { render, screen } from "@testing-library/react";
import Home from "@/app/page";

describe("Home", () => {
  it("renders the heading", () => {
    render(<Home />);
    expect(screen.getByRole("heading")).toBeInTheDocument();
  });
});
"#
}

fn eslint_config() -> &'static str {
    r#"import { dirname } from "path";
import { fileURLToPath } from "url";
import { FlatCompat } from "@eslint/eslintrc";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const compat = new FlatCompat({ baseDirectory: __dirname });

const eslintConfig = [
  ...compat.extends("next/core-web-vitals", "next/typescript"),
];

export default eslintConfig;
"#
}

fn env_example(cfg: &NextjsConfig) -> String {
    let template = include_str!("../templates/nextjs/env_example.rs");
    let mut db_url = String::new();
    let mut auth_vars = String::new();

    if cfg.orm != Orm::None {
        match cfg.db_provider {
            DbProvider::Postgres => {
                db_url.push_str("DATABASE_URL=postgresql://user:password@localhost:5432/mydb");
            }
            DbProvider::Mysql => {
                db_url.push_str("DATABASE_URL=mysql://user:password@localhost:3306/mydb");
            }
            DbProvider::Sqlite => {
                db_url.push_str("DATABASE_URL=dev.db");
            }
        }
    }

    if cfg.next_auth {
        auth_vars.push_str("AUTH_SECRET=changeme_generate_with_openssl_rand_base64_32\nAUTH_GITHUB_ID=\nAUTH_GITHUB_SECRET=");
    }

    if db_url.is_empty() && auth_vars.is_empty() {
        return "# Add your environment variables here\n".to_string();
    }

    template
        .replace("{{database_url}}", &db_url)
        .replace("{{auth_vars}}", &auth_vars)
}

fn nextjs_gitignore() -> &'static str {
    r#"# Dependencies
node_modules/
.pnp
.pnp.js

# Next.js
.next/
out/

# Production
build/
dist/

# Environment
.env
.env.local
.env.*.local

# Debug
npm-debug.log*
yarn-debug.log*
yarn-error.log*

# Misc
.DS_Store
*.pem
Thumbs.db

# Vercel
.vercel

# TypeScript
*.tsbuildinfo
next-env.d.ts
"#
}
